examples/jwt/verify-jwt.go

```diff
@@ -0,0 +1,187 @@
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Parse JWKs response.
+	var response jwksResponse
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return nil, err
+	}
+	if len(response.Keys) == 0 {
+		return nil, fmt.Errorf("no keys found")
+	}
+
+	// Construct a crypto.PublicKey from the response.
+	jwk := response.Keys[0]
+	n, err := base64.RawURLEncoding.DecodeString(jwk.N)
+	if err != nil {
+		return nil, err
+	}
+	e, err := base64.RawURLEncoding.DecodeString(jwk.E)
+	if err != nil {
+		return nil, err
+	}
+	return &rsa.PublicKey{
+		N: new(big.Int).SetBytes(n),
+		E: int(new(big.Int).SetBytes(e).Uint64()),
+	}, nil
+}
+
+// verify will verify the JWT.
+func verify(publicKey crypto.PublicKey, token string) (*claims, error) {
+	// Parse the raw token.
+	t, err := jwt.ParseSigned(token)
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate the signature on the JWT token.
+	var out claims
+	if err := t.Claims(publicKey, &out); err != nil {
+		return nil, err
+	}
+
+	return &out, nil
+}
+
+// validate validates the passed in claims against received claims.
+func validate(claims *claims, issuer string, subject string, audience string) error {
+	// Validate the claims on the JWT.
+	expectedClaims := jwt.Expected{
+		Issuer:   issuer,
+		Subject:  subject,
+		Audience: jwt.Audience{audience},
+		Time:     time.Now(),
+	}
+	if err := claims.Validate(expectedClaims); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	// Parse flags.
+	jwks := flag.String("jwks-url", "https://localhost:3080/.well-known/jwks.json", "JWK URL.")
+	skipVerify := flag.Bool("insecure-skip-verify", false, "Skip server certificate validation.")
+	jwt := flag.String("jwt", "", "JWT token to verify.")
+	validateClaims := flag.Bool("validate-claims", true, "Validate the claims received match expected.")
+	issuer := flag.String("issuer", "", "Issuer the name of the Teleport cluster.")
+	subject := flag.String("subject", "", "Subject is the identity of the Teleport user.")
+	audience := flag.String("audience", "", "Audience is the URI of the application.")
+	flag.Parse()
+
+	// Validate all required flags are set.
+	if *jwt == "" {
+		log.Fatal("JWT missing, required for validation.")
+	}
+	if *validateClaims {
+		if *issuer == "" || *subject == "" || *audience == "" {
+			log.Fatal("Issuer, Subject, and Audience required for validation.")
+		}
+	}
+
+	// Fetch and construct the public key that will be used to verify the JWT.
+	publicKey, err := getPublicKey(*jwks, *skipVerify)
+	if err != nil {
+		log.Fatalf("Failed to fetch JWKs needed to verify JWT: %v.", err)
+	}
+
+	// Verify the signature on the JWT.
+	claims, err := verify(publicKey, *jwt)
+	if err != nil {
+		log.Fatalf("JWT signature verification failed: %v.", err)
+	}
+
+	// Validate the claims if requested.
+	if *validateClaims {
+		if err := validate(claims, *issuer, *subject, *audience); err != nil {
+			log.Fatalf("Claim validation failed: %v.", err)
```

Comment:
I wonder if it would be helpful to still print the claims from the token in case of failed validation to help in troubleshooting.

Reply:
Done.